src/throttler/remote_thottler.js

```diff
@@ -0,0 +1,188 @@
+import Utils from '../util.js';
+
+const DEFAULT_REFRESH_INTERVAL_MS = 5000;
+const DEFAULT_INITIAL_DELAY_MS = 5000;
+const DEFAULT_THROTTLER_HOST = '0.0.0.0';
+const DEFAULT_THROTTLER_PORT = 5778;
+
+// minimumCredits is the minimum amount of credits necessary to not be throttled.
+// i.e. if currentCredits > MINIMUM_CREDITS, then the operation will not be throttled.
+const MINIMUM_CREDITS = 1.0;
+
+export default class RemoteThrottler {
+  _serviceName: string;
+  _logger: Logger;
+  _metrics: Metrics;
+
+  _refreshIntervalMs: number;
+  _host: string;
+  _port: number;
+
+  _uuid: string;
+  _credits: Map<string, number>;
+
+  _initialDelayTimeoutHandle: any;
+  _refreshIntervalHandle: any;
+  _onCreditsUpdate: ?Function;
+
+  /**
+   * Creates a RemoteThrottler that fetches credits remotely from jaeger-agent.
+   *
+   * @param {string} [serviceName] - name of the current service / application, same as given to Tracer
+   * @param {object} [options] - optional settings
+   * @param {object} [options.logger] - optional logger, see _flow/logger.js
+   * @param {object} [options.metrics] - instance of Metrics object
+   * @param {number} [options.refreshIntervalMs] - interval in milliseconds that determines how often credits
+   * are fetched (0 to not refresh, NOT RECOMMENDED!)
+   * @param {number} [options.initialDelayMs] - interval in milliseconds that determines how soon after initialization
+   * credits are first fetched (0 to not refresh, NOT RECOMMENDED!).
+   * @param {string} [options.host] - host for jaeger-agent, defaults to 'localhost'
+   * @param {number} [options.port] - port for jaeger-agent for /credits endpoint
+   * @param {function} [options.onCreditsUpdate] - callback function once credits are updated. Used for testing.
+   */
+  constructor(serviceName: string, options: any = {}) {
+    this._serviceName = serviceName;
+    this._logger = options.logger || new NullLogger();
+    this._metrics = options.metrics || new Metrics(new NoopMetricFactory());
+    this._refreshIntervalMs = options.refreshIntervalMs || DEFAULT_REFRESH_INTERVAL_MS;
+    this._host = options.host || DEFAULT_THROTTLER_HOST;
+    this._port = options.port || DEFAULT_THROTTLER_PORT;
+
+    this._credits = new Map();
+    this._onCreditsUpdate = options.onCreditsUpdate;
+
+    this._initialDelayTimeoutHandle = setTimeout(
+      this._afterInitialDelay.bind(this),
+      options.initialDelayMs || DEFAULT_INITIAL_DELAY_MS
+    );
+  }
+
+  _afterInitialDelay(): void {
+    this._refreshCredits();
+    this._refreshIntervalHandle = setInterval(this._refreshCredits.bind(this), this._refreshIntervalMs);
+  }
+
+  setProcess(process: Process): void {
+    this._uuid = process.uuid;
+  }
+
+  isAllowed(operation: string): boolean {
+    if (!this._credits.has(operation)) {
+      this._credits.set(operation, 0);
+      // If seen for the first time, async fetch credits
+      this._refreshCredits();
```

Comment:
In golang, when we see the operation for the first time, we have the option of allowing service owners to synchronously fetch credits for the first call; this is useful for cases like yab. In node however, I spoke with @tiffon and he said it's not advisable (TBF, it's not advisable in golang either but I think necessary).

I think this is an ok compromise.